src/proto/model.proto

```diff
@@ -0,0 +1,265 @@
+  // Set to true if we allow events to pass through without matching a column.
+  optional bool pass_through_non_matches = 3;
+}
+
+
+// Represents a diagonal matrix for updating event attributes.
+// We run the conditions in order, and once a condition matches, the update is
+// merged into the input event attributes, and we exit the run.
+message ConditionalMerge {
+  message ConditionalMergeNode {
+    // All conditions will be scanned sequentially until the first match is
+    // found.
+    optional FieldFilterProto condition = 1;
+
+    // What to update if the condition matches.
+    optional LabelerEvent update = 2;
+  }
+  // A list of conditions to go through.
+  repeated ConditionalMergeNode nodes = 1;
+  // Set to true if we allow events to not match any of the conditions.
+  optional bool pass_through_non_matches = 2;
+}
+
+
+// A matrix represented in a tree structure to reduce the column search space.
+// Implemented as an embedded model tree, represented by the root node.
+message UpdateTree {
+  optional CompiledNode root = 1;
+}
+
+
+message Multiplicity {
+  oneof multiplicity_ref {
+    // The constant expected multiplicity.
+    double expected_multiplicity = 1;
+    // The name of a field in LabelerEvent. Can be a nested field.
+    // Must be a singular signed numeric type.
+    // This is the reference to the field that contains the expected
+    // multiplicity value.
+    string expected_multiplicity_field = 2;
+  }
+  // The maximum allowed value stored in expected_multiplicity_field.
+  optional double max_value = 3;
+  // If true, cap the multiplicity value at max_value.
+  // Otherwise, return error if the value exceeds max_value.
+  optional bool cap_at_max = 4;
+  // The name of a field in LabelerEvent.
+  // Must be a singular signed int type field.
+  // This field is used as virtual person index at this multiplicity node.
+  optional string person_index_field = 5;
+}
+
+
+// Event and attributes produced by labeler. @labeler_input is the original
+// input event. @virtual_person_activity is the output label. All other entries
+// are inferred properties of the virtual people being generated, and are used
+// internally in the model.
+message LabelerEvent {
+  // Original input event.
+  optional LabelerInput labeler_input = 1;
+
+  // Labeler virtual person output. Including the VID, the demo of the virtual
+  // person, and other associated attributes.
+  // This is repeated because we allow multiple VIDs for single input event.
+  // For each event, there will always be 1 or more "virtual_person_activity"s.
+  // For each "virtual_person_activity", if virtual_person_id is not set, it is
+  // only for counting impressions by labels, and there is no virtual person for
+  // reach purpose.
+  repeated VirtualPersonActivity virtual_person_activities = 2;
+
+  // Sets to true if no user id is set for this event.
+  optional bool is_user_id_free_person = 3;
```

Comment:
input or output?  which field is for user_id?

Reply:
This field is used during applying the model, which is neither input nor output.

user_id is a field in the event schema, and will be added to event.proto in a future PR.